src/gen_llvm.cpp

```diff
@@ -3328,43 +3332,82 @@ SCOPES_RESULT(void) compile_object(const String *triple,
         LLVMCodeGenLevelDefault, LLVMRelocPIC, LLVMCodeModelJITDefault);
     assert(tm);
 
     char *path_cstr = strdup(path->data);
     LLVMBool failed = false;
-    switch(kind) {
-    case CFK_Object: {
-        failed = LLVMTargetMachineEmitToFile(tm, module, path_cstr,
-            LLVMObjectFile, &error_message);
-    } break;
-    case CFK_ASM: {
-        failed = LLVMTargetMachineEmitToFile(tm, module, path_cstr,
-            LLVMAssemblyFile, &error_message);
-    } break;
-    case CFK_BC: {
-        failed = LLVMWriteBitcodeToFile(module, path_cstr);
-    } break;
-    case CFK_LLVM: {
-        failed = LLVMPrintModuleToFile(module, path_cstr, &error_message);
-    } break;
-    default: {
+
+    if constexpr (std::is_void_v<T>) {
+        switch(kind) {
+            case CFK_Object: {
+                failed = LLVMTargetMachineEmitToFile(tm, module, path_cstr,
+                    LLVMObjectFile, &error_message);
+            } break;
+            case CFK_ASM: {
+                failed = LLVMTargetMachineEmitToFile(tm, module, path_cstr,
+                    LLVMAssemblyFile, &error_message);
+            } break;
+            case CFK_BC: {
+                failed = LLVMWriteBitcodeToFile(module, path_cstr);
+            } break;
+            case CFK_LLVM: {
+                failed = LLVMPrintModuleToFile(module, path_cstr, &error_message);
+            } break;
+            default: {
+                free(path_cstr);
+                SCOPES_ERROR(CGenBackendFailed, "unknown file kind");
+            } break;
+        }
+        
         free(path_cstr);
-        SCOPES_ERROR(CGenBackendFailed, "unknown file kind");
-    } break;
-    }
-    free(path_cstr);
-    if (failed) {
-        SCOPES_ERROR(CGenBackendFailed, error_message);
+        
+        if (failed) {
+            SCOPES_ERROR(CGenBackendFailed, error_message);
+        }
+        
+        return {};
+
+    } else if constexpr (std::is_same_v<T, const String*>) {
+        LLVMMemoryBufferRef buffer = nullptr;
+        LLVMBool failed = false;
+
+        switch(kind) {
+            case CFK_Object: {
+                failed = LLVMTargetMachineEmitToMemoryBuffer(tm, module, LLVMObjectFile, &error_message, &buffer);
+            } break;
+            case CFK_ASM: {
+                failed = LLVMTargetMachineEmitToMemoryBuffer(tm, module, LLVMAssemblyFile, &error_message, &buffer);
+            } break;
+            default: {
+                SCOPES_ERROR(CGenBackendFailed, "unknown file kind");
+            } break;
+        }
+
+        if (failed) {
+            SCOPES_ERROR(CGenBackendFailed, error_message);
+        }
+
+        const char* buffer_data = LLVMGetBufferStart(buffer);
+        const size_t buffer_size = LLVMGetBufferSize(buffer);
+
+        return String::from(buffer_data, buffer_size);
+    } else {
+        SCOPES_ERROR(CGenBackendFailed, "unimplemented template type");
```

Comment:
This should be a compile-type error, not a runtime error, because it applies to a compile time if expression. To achieve this, you'll need to use static_assert, but you can't simply do `static_assert(false, "unimplemented")` because some stupid C++ standard rules will make this fire no matter what. Instead, you'll need to make it dependent on the template type in a way that will always be false here, like `static_assert(std::is_void_v<T>, "unimplemented")`